crates/data_components/src/lib.rs

```diff
@@ -59,3 +64,25 @@ pub trait Stream {
         table_reference: OwnedTableReference,
     ) -> Result<Arc<dyn TableProvider + 'static>, Box<dyn Error + Send + Sync>>;
 }
+
+#[async_trait]
+pub trait DeleteTableProvider: TableProvider {
+    async fn delete(
+        &self,
+        _state: &SessionState,
+        _filters: &[Expr],
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        Err(DataFusionError::Plan("Not implemented".to_string()))
+    }
+}
+
+// There is no good way to allow inter trait casting yet as TableProvider is not controlled
+pub fn cast_to_deleteable<'a>(
```

Comment:
Would something like this not work?

https://github.com/datafusion-contrib/datafusion-federation/blob/main/datafusion-federation/src/analyzer.rs#L152

Reply:
they are same. Both are casting to a concrete implication, then use return type to cast back to dyn trait.